src/core/ext/filters/client_channel/lb_policy/xds/xds_override_host.cc

```diff
@@ -615,57 +633,96 @@ void XdsOverrideHostLb::Helper::AddTraceEvent(TraceSeverity severity,
 
 XdsOverrideHostLb::SubchannelWrapper::SubchannelWrapper(
     RefCountedPtr<SubchannelInterface> subchannel,
-    RefCountedPtr<XdsOverrideHostLb> policy,
-    absl::optional<const std::string> key)
+    RefCountedPtr<XdsOverrideHostLb> policy, absl::string_view key,
+    XdsHealthStatus health_status)
     : DelegatingSubchannel(std::move(subchannel)),
-      key_(std::move(key)),
-      policy_(std::move(policy)) {
-  auto watcher = std::make_unique<ConnectivityStateWatcher>();
+      key_(key),
+      policy_(policy),
+      health_status_(health_status) {
+  auto watcher = std::make_unique<ConnectivityStateWatcher>(Ref());
   watcher_ = watcher.get();
   wrapped_subchannel()->WatchConnectivityState(std::move(watcher));
 }
 
 XdsOverrideHostLb::SubchannelWrapper::~SubchannelWrapper() {
-  wrapped_subchannel()->CancelConnectivityStateWatch(watcher_);
   if (key_.has_value()) {
     policy_->ResetSubchannel(*key_, this);
   }
 }
 
 void XdsOverrideHostLb::SubchannelWrapper::WatchConnectivityState(
     std::unique_ptr<ConnectivityStateWatcherInterface> watcher) {
-  watcher_->watchers_.push_back(std::move(watcher));
+  watchers_.insert(std::move(watcher));
 }
 
 void XdsOverrideHostLb::SubchannelWrapper::CancelConnectivityStateWatch(
     ConnectivityStateWatcherInterface* watcher) {
-  (void)std::remove_if(watcher_->watchers_.begin(), watcher_->watchers_.end(),
-                       [watcher](const auto& watcher_ptr) {
-                         return watcher_ptr.get() == watcher;
-                       });
+  auto it = watchers_.find(watcher);
+  if (it != watchers_.end()) {
+    watchers_.erase(it);
+  }
+}
+
+void XdsOverrideHostLb::SubchannelWrapper::UpdateConnectivityState(
+    grpc_connectivity_state state, absl::Status status) {
+  connectivity_state_.store(state);
+  // Assumption is that watchers may come and go as the events are processed
```

Comment:
Hmm, that's interesting.  Can you point me to the tests that were failing, with log output?

In any case, if the issue here is that sending the connectivity state notification causes the child to change the set of watchers, then let's change the comment to be more specific about what the actual issue is here: "Sending connectivity state notifications to the watchers may cause the set of watchers to change, so we can't be iterating over the set of watchers while we send the notifications."

Reply:
Ah, okay -- it looks like this is happening during test shutdown.  The test shuts down the servers before shutting down the client, and when pick_first (which we use for LOGICAL_DNS clusters) sees the connection to the backend fail, it goes into IDLE state, at which point it stops watching the subchannel.

I guess this makes sense then.  It would be nice not to need it, but it's also not horrible.  I'll see if I can simplify this later via some structural changes to the `SubchannelInterface` API that I have in mind.